### PR TITLE
feat: add unit tests for error states in critical workflows

### PR DESCRIPTION
**Pull request by zwliew**
## Problem

#1110 added tests for the happy paths in critical workflows. It would be a good idea to add tests for the error paths as well to cover our bases.

Workflows:
1. Creating and sending a new email campaign
2. Creating and sending a new SMS campaign
3. Creating and sending a new Telegram campaign
4. Creating and sending a new protected email campaign
5. Viewing a protected email message
6. Unsubscribing from a mailing list

Partially resolves issue #1079 

## Solution

**Features**:

- Added unit tests for various error states that occur during campaign creation

## Before & After Screenshots

There are no visual changes.

## Tests

These tests have been verified on Travis and Amplify.

1. Deploy the change on Travis and Amplify.
2. All the following tests should be run and pass:
  - SMS
    - [x] Saving an invalid template (empty body after sanitization)
    - [x] Attempting to validate an invalid credential
    - [x] Uploading an invalid CSV file
   - Telegram
     - [x] Saving an invalid template (empty body after sanitization)
     - [x] Attempting to validate an invalid credential
     - [x] Uploading an invalid CSV file
   - Email
     - [x] Saving an invalid template (empty subject after sanitization)
     - [x] Uploading an invalid CSV file
   - Protected Email
     - [x] Saving an invalid template (empty subject after sanitization) (same as email)
     - [x] Saving an invalid template (extraneous subject params)
     - [x] Saving an invalid template (missing body params)
     - [x] Saving an invalid template (extraneous body params)
     - [x] Viewing a protected email with an invalid ID
     - [x] Viewing a protected email with an invalid password
     - [x] Uploading an invalid CSV file
- Write unit tests for the following misc non-error cases (not a priority; simply written for practice):
  - SMS
    - [x] Displays the necessary elements when rendered
    - [x] Next button is disabled when template is empty
    - [x] Next button is enabled when template is filled
    - [x] Character count text reflects the actual number of characters in the textbox
  - Telegram
    - [x] Displays the necessary elements when rendered
    - [x] Next button is disabled when template is empty
    - [x] Next button is enabled when template is filled
  - Email
    - [x] Displays the necessary elements when rendered

## Deploy Notes

The aforementioned tests have been added and should be run during every deployment on Travis and Amplify. If any test fails, the deployment should not go through.

Specifically, Amplify should not publish the new build if any test fails.

### CODE REVIEW

**Comment by zwliew** [comment]
I've rebased the branch on top of frontend-tests-dashboard

**Review by miazima — COMMENTED** [review]
Here are some suggestions on scenarios we could add, can be done in a later pr too

EmailRecipient 
- test for invalid file/valid file scenario

SMSCredential
- test with invalid phone number like `80000000`

File: frontend/src/components/dashboard/create/email/tests/EmailTemplate.test.tsx
Comment (zwliew) [reply frontend/src/components/dashboard/create/email/tests/EmailTemplate.test.tsx:69]: Thanks! I've added it

File: frontend/src/components/dashboard/create/sms/tests/SMSTemplate.test.tsx
Comment (zwliew) [reply frontend/src/components/dashboard/create/sms/tests/SMSTemplate.test.tsx:77]: Good point. I could create a common test for these 2 components, but I think maybe a cleaner option would be to extract the common code in `SMSTemplate` and `TelegramTemplate` into a separate component. I'll try that out in a separate PR.

File: frontend/src/components/dashboard/create/email/tests/EmailTemplate.test.tsx
Comment (zwliew) [reply frontend/src/components/dashboard/create/email/tests/EmailTemplate.test.tsx:104]: It's a bit difficult to do so, considering how each test has slightly different setup and teardown processes.

Comment (zwliew) [reply frontend/src/components/dashboard/create/email/tests/EmailTemplate.test.tsx:104]: I think it's a bit difficult to do so, since most tests have slightly different setup/teardown phases. For example, some tests mock `console.error` and some use protected email campaigns.

File: frontend/src/components/dashboard/create/sms/tests/SMSTemplate.test.tsx
Comment (zwliew) [reply frontend/src/components/dashboard/create/sms/tests/SMSTemplate.test.tsx:77]: I've attempted to extract the common code to a `BodyTemplate` component here: #1148

File: frontend/src/components/dashboard/create/email/tests/EmailTemplate.test.tsx
Comment (miazima) [reply frontend/src/components/dashboard/create/email/tests/EmailTemplate.test.tsx:104]: Alright, as for mocking the `console.error`, if you want to do this globally, you can feed it in a setup file in your `jest.config.ts`, this is something the backend tests currently implement to silence `console.log` messages

Comment (zwliew) [reply frontend/src/components/dashboard/create/email/tests/EmailTemplate.test.tsx:104]: Good point. I considered doing so, but decided against mocking `console.error` globally. When fixing test failures, I found that `console.error` statements provided useful information.

Comment (miazima) [reply frontend/src/components/dashboard/create/email/tests/EmailTemplate.test.tsx:104]: Sure, the backend tests currently implement that only for `console.log` to not overcrowd the travis log output

**Comment by zwliew** [comment]
Sorry for the erased commit history -- I've rebased the branch atop of `develop`. The latest changes start from commit 74b94ea, `feat: mock API responses for invalid CSV files`

File: frontend/src/components/dashboard/create/sms/tests/SMSTemplate.test.tsx
Comment (miazima) [reply frontend/src/components/dashboard/create/sms/tests/SMSTemplate.test.tsx:77]: Maybe this can go in first and then we can refactor the tests once the other pr is approved.

File: frontend/src/components/dashboard/create/email/tests/EmailTemplate.test.tsx
Comment (miazima) [reply frontend/src/components/dashboard/create/email/tests/EmailTemplate.test.tsx:97]: Do you want to add these in? i think the reason why I gave these examples is that they each output a different error message, but if the error messages are generated on the backend and not the frontend, then we can address this in backend tests

Comment (zwliew) [reply frontend/src/components/dashboard/create/email/tests/EmailTemplate.test.tsx:97]: I looked into them initially, but realised that I would need to add some extra cases to the API endpoint mocks to handle them, so I put that on hold 😅 

Now that you mention it, having these in the backend tests sounds like a good idea as well!

File: frontend/src/components/dashboard/create/sms/tests/SMSTemplate.test.tsx
Comment (zwliew) [reply frontend/src/components/dashboard/create/sms/tests/SMSTemplate.test.tsx:77]: Ok, sounds good!

**Review by miazima — APPROVED** [review]

**Pull request merged**
